{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7093eff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anndata\n",
    "\n",
    "from mgitools.os_helpers import listfiles"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73d62053",
   "metadata": {},
   "source": [
    "#### htan surgical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9fd83327",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "htan01 (2940, 33538)\n",
      "htan02 (1225, 33694)\n",
      "htan03 (666, 33694)\n",
      "htan04 (1346, 33694)\n",
      "htan05 (3462, 33538)\n",
      "htan06 (4145, 33538)\n"
     ]
    }
   ],
   "source": [
    "root = '../data/single_cell/original/inhouse/filtered_gene_bc_matrices/'\n",
    "\n",
    "sample_map = {\n",
    "    '1555-tumor': 'htan01',\n",
    "    'H_ZY-1105-07': 'htan02',\n",
    "    'H_ZY-1160-06': 'htan03',\n",
    "    'H_ZY-1174-06': 'htan04',\n",
    "    'TWCE-HT055P-S1PAG': 'htan05',\n",
    "    'TWCE-HT081P1-XB1': 'htan06'\n",
    "}\n",
    "\n",
    "adatas = []\n",
    "for sample, renamed_sample in sample_map.items():\n",
    "    fp = os.path.join(root, sample)\n",
    "    a = sc.read_10x_mtx(fp, var_names='gene_ids')\n",
    "    a.var_names_make_unique()\n",
    "    sc.pp.filter_cells(a, min_genes=100)\n",
    "    \n",
    "    a.obs['sample_id'] = renamed_sample\n",
    "    a.obs['dataset'] = 'htan_surgical'\n",
    "    a.obs.index = [f'{renamed_sample}_{x}' for x in a.obs.index]\n",
    "    print(renamed_sample, a.shape)\n",
    "    adatas.append(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ae758dd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbols</th>\n",
       "      <th>feature_types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277856</th>\n",
       "      <td>AC233755.2</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275063</th>\n",
       "      <td>AC233755.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271254</th>\n",
       "      <td>AC240274.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277475</th>\n",
       "      <td>AC213203.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268674</th>\n",
       "      <td>FAM231C</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33538 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_symbols    feature_types\n",
       "ENSG00000243485  MIR1302-2HG  Gene Expression\n",
       "ENSG00000237613      FAM138A  Gene Expression\n",
       "ENSG00000186092        OR4F5  Gene Expression\n",
       "ENSG00000238009   AL627309.1  Gene Expression\n",
       "ENSG00000239945   AL627309.3  Gene Expression\n",
       "...                      ...              ...\n",
       "ENSG00000277856   AC233755.2  Gene Expression\n",
       "ENSG00000275063   AC233755.1  Gene Expression\n",
       "ENSG00000271254   AC240274.1  Gene Expression\n",
       "ENSG00000277475   AC213203.1  Gene Expression\n",
       "ENSG00000268674      FAM231C  Gene Expression\n",
       "\n",
       "[33538 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adatas[0].var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "48a229f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGAGACCTAGG-1</th>\n",
       "      <td>1170</td>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGAGTGCGTGA-1</th>\n",
       "      <td>4305</td>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGCAATCGGTT-1</th>\n",
       "      <td>559</td>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGCACCGTTGG-1</th>\n",
       "      <td>1765</td>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGCATCCCACT-1</th>\n",
       "      <td>340</td>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_TTTGTCAGTCGCTTCT-1</th>\n",
       "      <td>411</td>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_TTTGTCAGTGACGGTA-1</th>\n",
       "      <td>877</td>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_TTTGTCAGTTGTACAC-1</th>\n",
       "      <td>3228</td>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_TTTGTCATCACTATTC-1</th>\n",
       "      <td>2133</td>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_TTTGTCATCCCAACGG-1</th>\n",
       "      <td>4111</td>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2940 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           n_genes sample_id        dataset\n",
       "htan01_AAACCTGAGACCTAGG-1     1170    htan01  htan_surgical\n",
       "htan01_AAACCTGAGTGCGTGA-1     4305    htan01  htan_surgical\n",
       "htan01_AAACCTGCAATCGGTT-1      559    htan01  htan_surgical\n",
       "htan01_AAACCTGCACCGTTGG-1     1765    htan01  htan_surgical\n",
       "htan01_AAACCTGCATCCCACT-1      340    htan01  htan_surgical\n",
       "...                            ...       ...            ...\n",
       "htan01_TTTGTCAGTCGCTTCT-1      411    htan01  htan_surgical\n",
       "htan01_TTTGTCAGTGACGGTA-1      877    htan01  htan_surgical\n",
       "htan01_TTTGTCAGTTGTACAC-1     3228    htan01  htan_surgical\n",
       "htan01_TTTGTCATCACTATTC-1     2133    htan01  htan_surgical\n",
       "htan01_TTTGTCATCCCAACGG-1     4111    htan01  htan_surgical\n",
       "\n",
       "[2940 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adatas[0].obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "181cbf52",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_id_to_symbol = {i:s\n",
    "                     for a in adatas\n",
    "                     for i, s in zip(a.var.index, a.var['gene_symbols'])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7856e1aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "for a in adatas:\n",
    "    a.obs = a.obs[['sample_id', 'dataset']]\n",
    "    # get rid of var so seurat conversion doesnt fail, adds back in later\n",
    "    a.var = a.var[[]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "50980715",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 13784 × 32991\n",
       "    obs: 'sample_id', 'dataset'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "htan_surgical = anndata.concat(adatas)\n",
    "htan_surgical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ef9a75d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGAGACCTAGG-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGAGTGCGTGA-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGCAATCGGTT-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGCACCGTTGG-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGCATCCCACT-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan06_TTTGGAGGTCAAACGG-1</th>\n",
       "      <td>htan06</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan06_TTTGGAGGTTGTCATG-1</th>\n",
       "      <td>htan06</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan06_TTTGGTTCATCCTAAG-1</th>\n",
       "      <td>htan06</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan06_TTTGGTTGTGCCCACA-1</th>\n",
       "      <td>htan06</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan06_TTTGTTGTCCTCTTTC-1</th>\n",
       "      <td>htan06</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13784 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          sample_id        dataset\n",
       "htan01_AAACCTGAGACCTAGG-1    htan01  htan_surgical\n",
       "htan01_AAACCTGAGTGCGTGA-1    htan01  htan_surgical\n",
       "htan01_AAACCTGCAATCGGTT-1    htan01  htan_surgical\n",
       "htan01_AAACCTGCACCGTTGG-1    htan01  htan_surgical\n",
       "htan01_AAACCTGCATCCCACT-1    htan01  htan_surgical\n",
       "...                             ...            ...\n",
       "htan06_TTTGGAGGTCAAACGG-1    htan06  htan_surgical\n",
       "htan06_TTTGGAGGTTGTCATG-1    htan06  htan_surgical\n",
       "htan06_TTTGGTTCATCCTAAG-1    htan06  htan_surgical\n",
       "htan06_TTTGGTTGTGCCCACA-1    htan06  htan_surgical\n",
       "htan06_TTTGTTGTCCTCTTTC-1    htan06  htan_surgical\n",
       "\n",
       "[13784 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "htan_surgical.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12ca827c",
   "metadata": {},
   "source": [
    "#### eus-fnb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e8cf6f81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wu01 (892, 33538)\n",
      "wu02 (1419, 33538)\n",
      "wu03 (1881, 33538)\n",
      "wu04 (141, 33538)\n",
      "wu05 (687, 33538)\n",
      "wu07 (3987, 33538)\n",
      "wu09 (211, 33538)\n",
      "wu14 (669, 33538)\n",
      "wu15 (1241, 33538)\n",
      "wu16 (3530, 33538)\n",
      "wu18 (1179, 33538)\n",
      "wu22 (755, 33538)\n",
      "wu23 (399, 33538)\n",
      "wu24 (408, 33538)\n",
      "wu25 (925, 33538)\n",
      "wu26 (2422, 33538)\n",
      "wu27 (2386, 33538)\n",
      "wu28 (2918, 33538)\n",
      "wu29 (1779, 33538)\n",
      "wu30 (2057, 33538)\n",
      "wu31 (1235, 33538)\n",
      "wu32 (973, 33538)\n",
      "wu38 (983, 33538)\n",
      "wu39 (9175, 33538)\n",
      "wu42 (1178, 33538)\n"
     ]
    }
   ],
   "source": [
    "keep = [\n",
    "    'wu01', 'wu02', 'wu03', 'wu04', 'wu05', 'wu07', 'wu09', 'wu14', 'wu15', 'wu16', 'wu18', 'wu22',\n",
    "    'wu23', 'wu24', 'wu25', 'wu26', 'wu27', 'wu28', 'wu29', 'wu30', 'wu31', 'wu32', 'wu38', 'wu39',\n",
    "    'wu42'\n",
    "]\n",
    "root = '../data/single_cell/original/inhouse/filtered_gene_bc_matrices/'\n",
    "\n",
    "adatas = []\n",
    "for sample in keep:\n",
    "    fp = os.path.join(root, sample)\n",
    "    a = sc.read_10x_mtx(fp, var_names='gene_ids')\n",
    "    a.var_names_make_unique()\n",
    "    sc.pp.filter_cells(a, min_genes=100)\n",
    "    a.obs['sample_id'] = sample\n",
    "    a.obs['dataset'] = 'eusfnb'\n",
    "    a.obs.index = [f'{sample}_{x}' for x in a.obs.index]\n",
    "    print(sample, a.shape)\n",
    "    adatas.append(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b4dd11e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "for a in adatas:\n",
    "    # in case not already in dict, add them\n",
    "    for gene_id, gene_symbol in zip(a.var.index.to_list(), a.var['gene_symbols']):\n",
    "        if gene_id not in gene_id_to_symbol:\n",
    "            gene_id_to_symbol[gene_id] = gene_symbol\n",
    "            \n",
    "    a.obs = a.obs[['sample_id', 'dataset']]\n",
    "    a.var = a.var[[]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a853823b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(43430, 33538)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eusfnb = anndata.concat(adatas)\n",
    "eusfnb.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fd259648",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>wu01_AAACCTGCACATAACC-1</th>\n",
       "      <td>wu01</td>\n",
       "      <td>eusfnb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wu01_AAACGGGCAAGGCTCC-1</th>\n",
       "      <td>wu01</td>\n",
       "      <td>eusfnb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wu01_AAACGGGTCATTCACT-1</th>\n",
       "      <td>wu01</td>\n",
       "      <td>eusfnb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wu01_AAACGGGTCTGTGCAA-1</th>\n",
       "      <td>wu01</td>\n",
       "      <td>eusfnb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wu01_AAAGTAGAGAACTGTA-1</th>\n",
       "      <td>wu01</td>\n",
       "      <td>eusfnb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wu42_TTTGGTTGTACCGTAT-1</th>\n",
       "      <td>wu42</td>\n",
       "      <td>eusfnb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wu42_TTTGGTTTCTACTTAC-1</th>\n",
       "      <td>wu42</td>\n",
       "      <td>eusfnb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wu42_TTTGGTTTCTCTGCTG-1</th>\n",
       "      <td>wu42</td>\n",
       "      <td>eusfnb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wu42_TTTGTCATCCTTGCCA-1</th>\n",
       "      <td>wu42</td>\n",
       "      <td>eusfnb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wu42_TTTGTCATCTGCAGTA-1</th>\n",
       "      <td>wu42</td>\n",
       "      <td>eusfnb</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43430 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        sample_id dataset\n",
       "wu01_AAACCTGCACATAACC-1      wu01  eusfnb\n",
       "wu01_AAACGGGCAAGGCTCC-1      wu01  eusfnb\n",
       "wu01_AAACGGGTCATTCACT-1      wu01  eusfnb\n",
       "wu01_AAACGGGTCTGTGCAA-1      wu01  eusfnb\n",
       "wu01_AAAGTAGAGAACTGTA-1      wu01  eusfnb\n",
       "...                           ...     ...\n",
       "wu42_TTTGGTTGTACCGTAT-1      wu42  eusfnb\n",
       "wu42_TTTGGTTTCTACTTAC-1      wu42  eusfnb\n",
       "wu42_TTTGGTTTCTCTGCTG-1      wu42  eusfnb\n",
       "wu42_TTTGTCATCCTTGCCA-1      wu42  eusfnb\n",
       "wu42_TTTGTCATCTGCAGTA-1      wu42  eusfnb\n",
       "\n",
       "[43430 rows x 2 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eusfnb.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e231240",
   "metadata": {},
   "source": [
    "#### peng"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e67ce04c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_map = {\n",
    "    'CRR034520': 'N1',\n",
    "    'CRR034521': 'N2',\n",
    "    'CRR034522': 'N3',\n",
    "    'CRR034523': 'N4',\n",
    "    'CRR034524': 'N5',\n",
    "    'CRR034525': 'N6',\n",
    "    'CRR034526': 'N7',\n",
    "    'CRR034527': 'N8',\n",
    "    'CRR034528': 'N9',\n",
    "    'CRR034529': 'N10',\n",
    "    'CRR034530': 'N11',\n",
    "    'CRR241805': 'T1',\n",
    "    'CRR241798': 'T2',\n",
    "    'CRR241799': 'T3',\n",
    "    'CRR034499': 'T4',\n",
    "    'CRR034500': 'T5',\n",
    "    'CRR034501': 'T6',\n",
    "    'CRR241800': 'T7',\n",
    "    'CRR034503': 'T8',\n",
    "    'CRR034504': 'T9',\n",
    "    'CRR034505': 'T10',\n",
    "    'CRR034506': 'T11',\n",
    "    'CRR034507': 'T12',\n",
    "    'CRR241801': 'T13',\n",
    "    'CRR034509': 'T14',\n",
    "    'CRR034510': 'T15',\n",
    "    'CRR034511': 'T16',\n",
    "    'CRR034512': 'T17',\n",
    "    'CRR034513': 'T18',\n",
    "    'CRR241802': 'T19',\n",
    "    'CRR241804': 'T20',\n",
    "    'CRR034516': 'T21',\n",
    "    'CRR034517': 'T22',\n",
    "    'CRR241803': 'T23',\n",
    "    'CRR034519': 'T24',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7669c8f3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N1 (9109, 33538)\n",
      "N2 (7498, 33538)\n",
      "N3 (3827, 33538)\n",
      "N4 (7918, 33538)\n",
      "N5 (4230, 33538)\n",
      "N6 (9286, 33538)\n",
      "N7 (6718, 33538)\n",
      "N8 (6682, 33538)\n",
      "N9 (9986, 33538)\n",
      "N10 (4293, 33538)\n",
      "N11 (11078, 33538)\n",
      "T1 (2882, 33538)\n",
      "T2 (8203, 33538)\n",
      "T3 (4401, 33538)\n",
      "T4 (3819, 33538)\n",
      "T5 (3640, 33538)\n",
      "T6 (4739, 33538)\n",
      "T7 (3581, 33538)\n",
      "T8 (3101, 33538)\n",
      "T9 (7224, 33538)\n",
      "T10 (2113, 33538)\n",
      "T11 (6954, 33538)\n",
      "T12 (5094, 33538)\n",
      "T13 (7567, 33538)\n",
      "T14 (4140, 33538)\n",
      "T15 (6939, 33538)\n",
      "T16 (4926, 33538)\n",
      "T17 (7885, 33538)\n",
      "T18 (6850, 33538)\n",
      "T19 (11108, 33538)\n",
      "T20 (4283, 33538)\n",
      "T21 (2792, 33538)\n",
      "T22 (5421, 33538)\n",
      "T23 (6947, 33538)\n",
      "T24 (9621, 33538)\n"
     ]
    }
   ],
   "source": [
    "root = '../data/single_cell/original/peng_reprocessed'\n",
    "\n",
    "adatas = []\n",
    "for sample, renamed_sample in sample_map.items():\n",
    "    fp = os.path.join(root, sample)\n",
    "    a = sc.read_10x_mtx(fp, var_names='gene_ids')\n",
    "    a.var_names_make_unique()\n",
    "    sc.pp.filter_cells(a, min_genes=100)\n",
    "    a.obs['sample_id'] = renamed_sample\n",
    "    a.obs['dataset'] = 'peng'\n",
    "    a.obs.index = [f'{renamed_sample}_{x}' for x in a.obs.index]\n",
    "    print(renamed_sample, a.shape)\n",
    "    adatas.append(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d38b6db1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for a in adatas:\n",
    "    # in case not already in dict, add them\n",
    "    for gene_id, gene_symbol in zip(a.var.index.to_list(), a.var['gene_symbols']):\n",
    "        if gene_id not in gene_id_to_symbol:\n",
    "            gene_id_to_symbol[gene_id] = gene_symbol\n",
    "            \n",
    "    a.obs = a.obs[['sample_id', 'dataset']]\n",
    "    a.var = a.var[[]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c2e02f79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(214855, 33538)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peng = anndata.concat(adatas)\n",
    "peng.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2d013296",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>N1_AAACCTGAGAACAATC-1</th>\n",
       "      <td>N1</td>\n",
       "      <td>peng</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N1_AAACCTGAGCTATGCT-1</th>\n",
       "      <td>N1</td>\n",
       "      <td>peng</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N1_AAACCTGAGGTGTTAA-1</th>\n",
       "      <td>N1</td>\n",
       "      <td>peng</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N1_AAACCTGAGTATGACA-1</th>\n",
       "      <td>N1</td>\n",
       "      <td>peng</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N1_AAACCTGCAACCGCCA-1</th>\n",
       "      <td>N1</td>\n",
       "      <td>peng</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T24_TTTGTCAGTCTAAAGA-1</th>\n",
       "      <td>T24</td>\n",
       "      <td>peng</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T24_TTTGTCATCAGTTTGG-1</th>\n",
       "      <td>T24</td>\n",
       "      <td>peng</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T24_TTTGTCATCCCATTTA-1</th>\n",
       "      <td>T24</td>\n",
       "      <td>peng</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T24_TTTGTCATCCTTTCTC-1</th>\n",
       "      <td>T24</td>\n",
       "      <td>peng</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T24_TTTGTCATCGAATGCT-1</th>\n",
       "      <td>T24</td>\n",
       "      <td>peng</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>214855 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       sample_id dataset\n",
       "N1_AAACCTGAGAACAATC-1         N1    peng\n",
       "N1_AAACCTGAGCTATGCT-1         N1    peng\n",
       "N1_AAACCTGAGGTGTTAA-1         N1    peng\n",
       "N1_AAACCTGAGTATGACA-1         N1    peng\n",
       "N1_AAACCTGCAACCGCCA-1         N1    peng\n",
       "...                          ...     ...\n",
       "T24_TTTGTCAGTCTAAAGA-1       T24    peng\n",
       "T24_TTTGTCATCAGTTTGG-1       T24    peng\n",
       "T24_TTTGTCATCCCATTTA-1       T24    peng\n",
       "T24_TTTGTCATCCTTTCTC-1       T24    peng\n",
       "T24_TTTGTCATCGAATGCT-1       T24    peng\n",
       "\n",
       "[214855 rows x 2 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peng.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "969c83f9",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "###### old (not used anymore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b1fd3e6e",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# peng = sc.read_h5ad('../data/single_cell/original/peng/PRJCA001063_CRC_besca2.raw.h5ad')\n",
    "# # peng.var_names_make_unique()\n",
    "# # peng.obs_names_make_unique()\n",
    "# peng"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2c2e92ca",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# peng.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9a0df94e",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# peng.obs['sample_id'] = peng.obs['Patient'].to_list()\n",
    "# peng.obs['dataset'] = 'peng'\n",
    "# peng.obs = peng.obs[['sample_id', 'dataset']]\n",
    "# peng.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c84affd4",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# peng.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "25f35b17",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# for gene_id, gene_symbol in zip(peng.var['ENSEMBL'], peng.var['SYMBOL']):\n",
    "#     if gene_id not in gene_id_to_symbol:\n",
    "#         gene_id_to_symbol[gene_id] = gene_symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b06e9bb5",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# peng.var.index = peng.var['ENSEMBL'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b6aaadfa",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# pool = set(peng.var.index)\n",
    "# len(pool)\n",
    "# len(pool.intersection(set(eusfnb.var.index)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c9437149",
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# peng.var = peng.var[[]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3fee2a6",
   "metadata": {},
   "source": [
    "#### lin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b8e0a39f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_lin(matrix_fp, features_fp, barcodes_fp):\n",
    "    mat = sc.read_mtx(matrix_fp).X.transpose()\n",
    "    feats = pd.read_csv(features_fp, sep='\\t', header=None, index_col=0)\n",
    "    feats.index.name = 'gene_ids'\n",
    "    ls = list(feats.columns)\n",
    "    ls[0] = 'gene_symbols'\n",
    "    feats.columns = ls\n",
    "    barcodes = pd.read_csv(barcodes_fp, sep='\\t', header=None, index_col=0)\n",
    "    a = anndata.AnnData(X=mat, obs=barcodes, var=feats)\n",
    "    return a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "fe72ae8c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MET01': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/MET01/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/MET01/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/MET01/matrix.mtx.gz'},\n",
       " 'MET02': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/MET02/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/MET02/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/MET02/matrix.mtx.gz'},\n",
       " 'MET03': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/MET03/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/MET03/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/MET03/matrix.mtx.gz'},\n",
       " 'MET04': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/MET04/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/MET04/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/MET04/matrix.mtx.gz'},\n",
       " 'MET05': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/MET05/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/MET05/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/MET05/matrix.mtx.gz'},\n",
       " 'MET06': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/MET06/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/MET06/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/MET06/matrix.mtx.gz'},\n",
       " 'P01': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/P01/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/P01/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/P01/matrix.mtx.gz'},\n",
       " 'P02': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/P02/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/P02/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/P02/matrix.mtx.gz'},\n",
       " 'P03': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/P03/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/P03/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/P03/matrix.mtx.gz'},\n",
       " 'P04': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/P04/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/P04/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/P04/matrix.mtx.gz'},\n",
       " 'P05': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/P05/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/P05/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/P05/matrix.mtx.gz'},\n",
       " 'P06': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/P06/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/P06/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/P06/matrix.mtx.gz'},\n",
       " 'P07': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/P07/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/P07/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/P07/matrix.mtx.gz'},\n",
       " 'P08': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/P08/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/P08/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/P08/matrix.mtx.gz'},\n",
       " 'P09': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/P09/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/P09/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/P09/matrix.mtx.gz'},\n",
       " 'P10': {'barcodes': '../data/single_cell/original/lin/GSE154778_RAW/P10/barcodes.tsv.gz',\n",
       "  'features': '../data/single_cell/original/lin/GSE154778_RAW/P10/features.tsv.gz',\n",
       "  'matrix': '../data/single_cell/original/lin/GSE154778_RAW/P10/matrix.mtx.gz'}}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fps = sorted(listfiles('../data/single_cell/original/lin/GSE154778_RAW', regex=r'.gz$'))\n",
    "m = {}\n",
    "for fp in fps:\n",
    "    sample = fp.split('/')[-2]\n",
    "    if sample not in m:\n",
    "        m[sample] = {}\n",
    "    if 'barcodes' in fp:\n",
    "        m[sample]['barcodes'] = fp\n",
    "    elif 'features' in fp:\n",
    "        m[sample]['features'] = fp\n",
    "    elif 'matrix' in fp:\n",
    "        m[sample]['matrix'] = fp\n",
    "m       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "cbd643d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MET01\n",
      "MET02\n",
      "MET03\n",
      "MET04\n",
      "MET05\n",
      "MET06\n",
      "P01\n",
      "P02\n",
      "P03\n",
      "P04\n",
      "P05\n",
      "P06\n",
      "P07\n",
      "P08\n",
      "P09\n",
      "P10\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adatas = []\n",
    "for sample, d in m.items():\n",
    "    print(sample)\n",
    "    a = read_lin(d['matrix'], d['features'], d['barcodes'])\n",
    "    a.obs['sample_id'] = sample\n",
    "    a.obs.index = [f'{sample}_{x}' for x in a.obs.index.to_list()]\n",
    "    adatas.append(a)\n",
    "len(adatas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "69aff72d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for a in adatas:\n",
    "    for gene_id, gene_symbol in zip(a.var.index.to_list(), a.var['gene_symbols']):\n",
    "        if gene_id not in gene_id_to_symbol:\n",
    "            gene_id_to_symbol[gene_id] = gene_symbol\n",
    "    a.var = a.var[[]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2db0c860",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 17086 × 30358\n",
       "    obs: 'sample_id'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lin = anndata.concat(adatas)\n",
    "lin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1c344bf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "lin.obs['dataset'] = 'lin'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "efac4f3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MET01_AAAGATGAGCCCGAAA-1</th>\n",
       "      <td>MET01</td>\n",
       "      <td>lin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MET01_AAAGATGAGTCAAGGC-1</th>\n",
       "      <td>MET01</td>\n",
       "      <td>lin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MET01_AAAGATGCAGCGAACA-1</th>\n",
       "      <td>MET01</td>\n",
       "      <td>lin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MET01_AAAGTAGGTACCTACA-1</th>\n",
       "      <td>MET01</td>\n",
       "      <td>lin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MET01_AACTCAGAGCAGCGTA-1</th>\n",
       "      <td>MET01</td>\n",
       "      <td>lin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P10_TTTGTCAGTTGATTGC-1</th>\n",
       "      <td>P10</td>\n",
       "      <td>lin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P10_TTTGTCATCACTGGGC-1</th>\n",
       "      <td>P10</td>\n",
       "      <td>lin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P10_TTTGTCATCAGAGACG-1</th>\n",
       "      <td>P10</td>\n",
       "      <td>lin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P10_TTTGTCATCATGGTCA-1</th>\n",
       "      <td>P10</td>\n",
       "      <td>lin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P10_TTTGTCATCCGTTGTC-1</th>\n",
       "      <td>P10</td>\n",
       "      <td>lin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17086 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         sample_id dataset\n",
       "MET01_AAAGATGAGCCCGAAA-1     MET01     lin\n",
       "MET01_AAAGATGAGTCAAGGC-1     MET01     lin\n",
       "MET01_AAAGATGCAGCGAACA-1     MET01     lin\n",
       "MET01_AAAGTAGGTACCTACA-1     MET01     lin\n",
       "MET01_AACTCAGAGCAGCGTA-1     MET01     lin\n",
       "...                            ...     ...\n",
       "P10_TTTGTCAGTTGATTGC-1         P10     lin\n",
       "P10_TTTGTCATCACTGGGC-1         P10     lin\n",
       "P10_TTTGTCATCAGAGACG-1         P10     lin\n",
       "P10_TTTGTCATCATGGTCA-1         P10     lin\n",
       "P10_TTTGTCATCCGTTGTC-1         P10     lin\n",
       "\n",
       "[17086 rows x 2 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lin.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3247b94c",
   "metadata": {},
   "source": [
    "#### chan-seng-yue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e4a4a8b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/single_cell/original/chan-seng-yue/h5/100070_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/85948_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/87235_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/87784_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/90209_CMP_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/91412_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/91610_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/91706_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/94930_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/95092_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/95373_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/96460_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/97727_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/COMP_0158_P_filtered_gene_bc_matrices_h5.h5',\n",
       " '../data/single_cell/original/chan-seng-yue/h5/G9903_filtered_gene_bc_matrices_h5.h5']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fps = sorted(listfiles('../data/single_cell/original/chan-seng-yue/h5', regex=r'.h5$'))\n",
    "fps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b9dadb4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/diskmnt/Projects/Users/estorrs/miniconda3/envs/single-cell-pdac/lib/python3.7/site-packages/anndata/_core/anndata.py:1830: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100070 (2789, 32738)\n",
      "85948 (2323, 32738)\n",
      "87235 (2718, 32738)\n",
      "87784 (2325, 32738)\n",
      "90209_CMP (1641, 32738)\n",
      "91412 (2347, 32738)\n",
      "91610 (1250, 32738)\n",
      "91706 (2662, 32738)\n",
      "94930 (1428, 32738)\n",
      "95092 (2567, 32738)\n",
      "95373 (2362, 32738)\n",
      "96460 (1260, 32738)\n",
      "97727 (2657, 32738)\n",
      "COMP_0158_P (3159, 32738)\n",
      "G9903 (2477, 32738)\n"
     ]
    }
   ],
   "source": [
    "adatas = []\n",
    "for fp in fps:\n",
    "    sample = fp.split('/')[-1].split('_filtered')[0]\n",
    "    a = sc.read_10x_h5(fp)\n",
    "    a.var_names_make_unique()\n",
    "    sc.pp.filter_cells(a, min_genes=100)\n",
    "    a.obs['sample_id'] = sample\n",
    "    a.obs['dataset'] = 'chan-seng-yue'\n",
    "    a.obs.index = [f'{sample}_{x}' for x in a.obs.index]\n",
    "    print(sample, a.shape)\n",
    "    adatas.append(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "66587029",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-10</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.7</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.8</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC145205.1</th>\n",
       "      <td>ENSG00000215635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BAGE5</th>\n",
       "      <td>ENSG00000268590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CU459201.1</th>\n",
       "      <td>ENSG00000251180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC002321.2</th>\n",
       "      <td>ENSG00000215616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC002321.1</th>\n",
       "      <td>ENSG00000215611</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32738 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene_ids\n",
       "MIR1302-10    ENSG00000243485\n",
       "FAM138A       ENSG00000237613\n",
       "OR4F5         ENSG00000186092\n",
       "RP11-34P13.7  ENSG00000238009\n",
       "RP11-34P13.8  ENSG00000239945\n",
       "...                       ...\n",
       "AC145205.1    ENSG00000215635\n",
       "BAGE5         ENSG00000268590\n",
       "CU459201.1    ENSG00000251180\n",
       "AC002321.2    ENSG00000215616\n",
       "AC002321.1    ENSG00000215611\n",
       "\n",
       "[32738 rows x 1 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adatas[0].var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9fe765ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "for a in adatas:\n",
    "    for gene_id, gene_symbol in zip(a.var['gene_ids'], a.var.index.to_list()):\n",
    "        if gene_id not in gene_id_to_symbol:\n",
    "            gene_id_to_symbol[gene_id] = gene_symbol\n",
    "    a.obs = a.obs[['sample_id', 'dataset']]\n",
    "    a.var.index = a.var['gene_ids'].to_list()\n",
    "    a.var = a.var[[]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "65a30e2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33965, 32738)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csy = anndata.concat(adatas)\n",
    "csy.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "43e7197c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>100070_AAACCTGAGCGTAATA-1</th>\n",
       "      <td>100070</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100070_AAACCTGCAATCCGAT-1</th>\n",
       "      <td>100070</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100070_AAACCTGCACCGATAT-1</th>\n",
       "      <td>100070</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100070_AAACCTGCACTAGTAC-1</th>\n",
       "      <td>100070</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100070_AAACGGGAGACTGGGT-1</th>\n",
       "      <td>100070</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCAAGTTGTCGT-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCACAACTTGAC-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCACATATGGTC-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCATCCGAACGC-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCATCTCTTGAT-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33965 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          sample_id        dataset\n",
       "100070_AAACCTGAGCGTAATA-1    100070  chan-seng-yue\n",
       "100070_AAACCTGCAATCCGAT-1    100070  chan-seng-yue\n",
       "100070_AAACCTGCACCGATAT-1    100070  chan-seng-yue\n",
       "100070_AAACCTGCACTAGTAC-1    100070  chan-seng-yue\n",
       "100070_AAACGGGAGACTGGGT-1    100070  chan-seng-yue\n",
       "...                             ...            ...\n",
       "G9903_TTTGTCAAGTTGTCGT-1      G9903  chan-seng-yue\n",
       "G9903_TTTGTCACAACTTGAC-1      G9903  chan-seng-yue\n",
       "G9903_TTTGTCACATATGGTC-1      G9903  chan-seng-yue\n",
       "G9903_TTTGTCATCCGAACGC-1      G9903  chan-seng-yue\n",
       "G9903_TTTGTCATCTCTTGAT-1      G9903  chan-seng-yue\n",
       "\n",
       "[33965 rows x 2 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csy.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bcf130a",
   "metadata": {},
   "source": [
    "#### combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "a2046bab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 323120 × 35476\n",
       "    obs: 'sample_id', 'dataset'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = anndata.concat([htan_surgical, eusfnb, lin, peng, csy], join='outer')\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "bf0b75cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbols</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>TSPAN6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>TNMD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>DPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>SCYL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>C1orf112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285492</th>\n",
       "      <td>AL356417.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285505</th>\n",
       "      <td>AC010616.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285508</th>\n",
       "      <td>AL034430.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285509</th>\n",
       "      <td>AP000646.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285513</th>\n",
       "      <td>AP006216.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35476 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_symbols\n",
       "ENSG00000000003       TSPAN6\n",
       "ENSG00000000005         TNMD\n",
       "ENSG00000000419         DPM1\n",
       "ENSG00000000457        SCYL3\n",
       "ENSG00000000460     C1orf112\n",
       "...                      ...\n",
       "ENSG00000285492   AL356417.3\n",
       "ENSG00000285505   AC010616.2\n",
       "ENSG00000285508   AL034430.1\n",
       "ENSG00000285509   AP000646.1\n",
       "ENSG00000285513   AP006216.3\n",
       "\n",
       "[35476 rows x 1 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var['gene_symbols'] = [gene_id_to_symbol[gid] for gid in adata.var.index.to_list()]\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "4caeed27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "34985"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(adata.var['gene_symbols']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "c45c933f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'100070',\n",
       " '85948',\n",
       " '87235',\n",
       " '87784',\n",
       " '90209_CMP',\n",
       " '91412',\n",
       " '91610',\n",
       " '91706',\n",
       " '94930',\n",
       " '95092',\n",
       " '95373',\n",
       " '96460',\n",
       " '97727',\n",
       " 'COMP_0158_P',\n",
       " 'G9903',\n",
       " 'MET01',\n",
       " 'MET02',\n",
       " 'MET03',\n",
       " 'MET04',\n",
       " 'MET05',\n",
       " 'MET06',\n",
       " 'N1',\n",
       " 'N10',\n",
       " 'N11',\n",
       " 'N2',\n",
       " 'N3',\n",
       " 'N4',\n",
       " 'N5',\n",
       " 'N6',\n",
       " 'N7',\n",
       " 'N8',\n",
       " 'N9',\n",
       " 'P01',\n",
       " 'P02',\n",
       " 'P03',\n",
       " 'P04',\n",
       " 'P05',\n",
       " 'P06',\n",
       " 'P07',\n",
       " 'P08',\n",
       " 'P09',\n",
       " 'P10',\n",
       " 'T1',\n",
       " 'T10',\n",
       " 'T11',\n",
       " 'T12',\n",
       " 'T13',\n",
       " 'T14',\n",
       " 'T15',\n",
       " 'T16',\n",
       " 'T17',\n",
       " 'T18',\n",
       " 'T19',\n",
       " 'T2',\n",
       " 'T20',\n",
       " 'T21',\n",
       " 'T22',\n",
       " 'T23',\n",
       " 'T24',\n",
       " 'T3',\n",
       " 'T4',\n",
       " 'T5',\n",
       " 'T6',\n",
       " 'T7',\n",
       " 'T8',\n",
       " 'T9',\n",
       " 'htan01',\n",
       " 'htan02',\n",
       " 'htan03',\n",
       " 'htan04',\n",
       " 'htan05',\n",
       " 'htan06',\n",
       " 'wu01',\n",
       " 'wu02',\n",
       " 'wu03',\n",
       " 'wu04',\n",
       " 'wu05',\n",
       " 'wu07',\n",
       " 'wu09',\n",
       " 'wu14',\n",
       " 'wu15',\n",
       " 'wu16',\n",
       " 'wu18',\n",
       " 'wu22',\n",
       " 'wu23',\n",
       " 'wu24',\n",
       " 'wu25',\n",
       " 'wu26',\n",
       " 'wu27',\n",
       " 'wu28',\n",
       " 'wu29',\n",
       " 'wu30',\n",
       " 'wu31',\n",
       " 'wu32',\n",
       " 'wu38',\n",
       " 'wu39',\n",
       " 'wu42'}"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(adata.obs['sample_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "18f1c8de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGAGACCTAGG-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGAGTGCGTGA-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGCAATCGGTT-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGCACCGTTGG-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGCATCCCACT-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCAAGTTGTCGT-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCACAACTTGAC-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCACATATGGTC-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCATCCGAACGC-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCATCTCTTGAT-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>323120 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          sample_id        dataset\n",
       "htan01_AAACCTGAGACCTAGG-1    htan01  htan_surgical\n",
       "htan01_AAACCTGAGTGCGTGA-1    htan01  htan_surgical\n",
       "htan01_AAACCTGCAATCGGTT-1    htan01  htan_surgical\n",
       "htan01_AAACCTGCACCGTTGG-1    htan01  htan_surgical\n",
       "htan01_AAACCTGCATCCCACT-1    htan01  htan_surgical\n",
       "...                             ...            ...\n",
       "G9903_TTTGTCAAGTTGTCGT-1      G9903  chan-seng-yue\n",
       "G9903_TTTGTCACAACTTGAC-1      G9903  chan-seng-yue\n",
       "G9903_TTTGTCACATATGGTC-1      G9903  chan-seng-yue\n",
       "G9903_TTTGTCATCCGAACGC-1      G9903  chan-seng-yue\n",
       "G9903_TTTGTCATCTCTTGAT-1      G9903  chan-seng-yue\n",
       "\n",
       "[323120 rows x 2 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e6f66967",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>dataset</th>\n",
       "      <th>tissue_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGAGACCTAGG-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "      <td>tumor_primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGAGTGCGTGA-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "      <td>tumor_primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGCAATCGGTT-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "      <td>tumor_primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGCACCGTTGG-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "      <td>tumor_primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>htan01_AAACCTGCATCCCACT-1</th>\n",
       "      <td>htan01</td>\n",
       "      <td>htan_surgical</td>\n",
       "      <td>tumor_primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCAAGTTGTCGT-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "      <td>tumor_primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCACAACTTGAC-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "      <td>tumor_primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCACATATGGTC-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "      <td>tumor_primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCATCCGAACGC-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "      <td>tumor_primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_TTTGTCATCTCTTGAT-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>chan-seng-yue</td>\n",
       "      <td>tumor_primary</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>323120 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          sample_id        dataset    tissue_type\n",
       "htan01_AAACCTGAGACCTAGG-1    htan01  htan_surgical  tumor_primary\n",
       "htan01_AAACCTGAGTGCGTGA-1    htan01  htan_surgical  tumor_primary\n",
       "htan01_AAACCTGCAATCGGTT-1    htan01  htan_surgical  tumor_primary\n",
       "htan01_AAACCTGCACCGTTGG-1    htan01  htan_surgical  tumor_primary\n",
       "htan01_AAACCTGCATCCCACT-1    htan01  htan_surgical  tumor_primary\n",
       "...                             ...            ...            ...\n",
       "G9903_TTTGTCAAGTTGTCGT-1      G9903  chan-seng-yue  tumor_primary\n",
       "G9903_TTTGTCACAACTTGAC-1      G9903  chan-seng-yue  tumor_primary\n",
       "G9903_TTTGTCACATATGGTC-1      G9903  chan-seng-yue  tumor_primary\n",
       "G9903_TTTGTCATCCGAACGC-1      G9903  chan-seng-yue  tumor_primary\n",
       "G9903_TTTGTCATCTCTTGAT-1      G9903  chan-seng-yue  tumor_primary\n",
       "\n",
       "[323120 rows x 3 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_tissue_type(sample, dataset):\n",
    "    if dataset == 'peng' and sample[0] == 'N':\n",
    "        return 'normal'\n",
    "    if dataset == 'lin' and 'MET' in sample:\n",
    "        return 'tumor_met'\n",
    "    return 'tumor_primary'\n",
    "\n",
    "adata.obs['tissue_type'] = [get_tissue_type(s, d)\n",
    "                            for s, d in zip(adata.obs['sample_id'], adata.obs['dataset'])]\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "c2ff3996",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('tumor_primary', 235030), ('normal', 80625), ('tumor_met', 7465)]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "Counter(adata.obs['tissue_type']).most_common()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "c4e066dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad('../data/sandbox/temp.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92d92d49",
   "metadata": {},
   "source": [
    "###### build count matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "bcc6bf8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(323120, 35476)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad('../data/sandbox/temp.h5ad')\n",
    "adata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "75749886",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbols</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL356417.3</th>\n",
       "      <td>AL356417.3</td>\n",
       "      <td>ENSG00000285492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC010616.2</th>\n",
       "      <td>AC010616.2</td>\n",
       "      <td>ENSG00000285505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL034430.1</th>\n",
       "      <td>AL034430.1</td>\n",
       "      <td>ENSG00000285508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AP000646.1</th>\n",
       "      <td>AP000646.1</td>\n",
       "      <td>ENSG00000285509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AP006216.3</th>\n",
       "      <td>AP006216.3</td>\n",
       "      <td>ENSG00000285513</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35476 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_symbols          gene_id\n",
       "TSPAN6           TSPAN6  ENSG00000000003\n",
       "TNMD               TNMD  ENSG00000000005\n",
       "DPM1               DPM1  ENSG00000000419\n",
       "SCYL3             SCYL3  ENSG00000000457\n",
       "C1orf112       C1orf112  ENSG00000000460\n",
       "...                 ...              ...\n",
       "AL356417.3   AL356417.3  ENSG00000285492\n",
       "AC010616.2   AC010616.2  ENSG00000285505\n",
       "AL034430.1   AL034430.1  ENSG00000285508\n",
       "AP000646.1   AP000646.1  ENSG00000285509\n",
       "AP006216.3   AP006216.3  ENSG00000285513\n",
       "\n",
       "[35476 rows x 2 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var['gene_id'] = adata.var.index.to_list()\n",
    "adata.var.index = adata.var['gene_symbols'].to_list()\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "67d45390",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var = adata.var[[]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "cbbb78c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.index.name = 'cell_id'\n",
    "adata.obs.to_csv('../data/single_cell/original/count_matrix/metadata.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "dbbe9620",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "34985"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(adata.var.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "0b853995",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL356417.3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC010616.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL034430.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AP000646.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AP006216.3</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35476 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [TSPAN6, TNMD, DPM1, SCYL3, C1orf112, FGR, CFH, FUCA2, GCLC, NFYA, STPG1, NIPAL3, LAS1L, ENPP4, SEMA3F, CFTR, ANKIB1, CYP51A1, KRIT1, RAD52, BAD, LAP3, CD99, HS3ST1, AOC1, WNT16, HECW1, MAD1L1, LASP1, SNX11, TMEM176A, M6PR, KLHL13, CYP26B1, ICA1, DBNDD1, ALS2, CASP10, CFLAR, TFPI, NDUFAF7, RBM5, MTMR7, SLC7A2, ARF5, SARM1, POLDIP2, PLXND1, AK2, CD38, FKBP4, KDM1A, RBM6, CAMKK1, RECQL, VPS50, HSPB6, ARHGAP33, NDUFAB1, PDK4, SLC22A16, ZMYND10, ABCB5, ARX, SLC25A13, ST7, CDC27, SLC4A1, CALCR, HCCS, DVL2, PRSS22, UPF1, SKAP2, SLC25A5, MCUB, HOXA11, POLR2J, DHX33, MEOX1, THSD7A, LIG3, RPAP3, ACSM3, REXO5, CIAPIN1, SPPL2B, FAM214B, COPZ2, PRKAR2B, MSL3, CREBBP, TSPOAP1, MPO, PON1, GCFC2, WDR54, CROT, ABCB4, KMT2E, ...]\n",
       "\n",
       "[35476 rows x 0 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "374e23a0",
   "metadata": {},
   "source": [
    "keep only first entry if there are multiple gene ids mapping to single gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "7083efcc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(34985, 34985)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seen = set()\n",
    "keep = []\n",
    "for i, g in enumerate(adata.var.index.to_list()):\n",
    "    if g not in seen:\n",
    "        seen.add(g)\n",
    "        keep.append(i)\n",
    "len(keep), len(seen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "ee9d86c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(323120, 34985)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(adata.obs.index)), len(set(adata.var.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "abc9b2c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 323120 × 34985\n",
       "    obs: 'sample_id', 'dataset', 'tissue_type'"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slim = adata[:, keep]\n",
    "slim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "b210023f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "34985"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(slim.var.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "ff8a49ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "slim.write_h5ad('../data/single_cell/original/count_matrix/counts.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03b1beb0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26810835",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6758284e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e3aa73a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 140612 × 30840\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'nCount_SCT', 'nFeature_SCT', 'CELL', 'CONDITION', 'Patient', 'Type', 'Cell_type', 'integrated_snn_res.1.25', 'seurat_clusters', 'integrated_snn_res.2', 'integrated_snn_res.3', 'integrated_snn_res.1', 'integrated_snn_res.0.5', 'active.ident', 'cell_type', 'dataset', 'sample_id', 'cell_type_level_2', 'cell_type_level_3', 'tumor_original_seurat', 'malignant_exclude', 'final_cell_types'\n",
       "    var: 'is_highly_variable'\n",
       "    uns: 'cell_type_colors', 'cell_type_level_2_colors', 'cell_type_level_3_colors', 'dataset_colors', 'final_cell_types_colors', 'rank_genes_groups'\n",
       "    obsm: 'X_umap'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = sc.read_h5ad('../../sc_pdac/backups/final_adata.h5ad')\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6f369e7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>nCount_SCT</th>\n",
       "      <th>nFeature_SCT</th>\n",
       "      <th>CELL</th>\n",
       "      <th>CONDITION</th>\n",
       "      <th>Patient</th>\n",
       "      <th>Type</th>\n",
       "      <th>Cell_type</th>\n",
       "      <th>...</th>\n",
       "      <th>integrated_snn_res.0.5</th>\n",
       "      <th>active.ident</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>dataset</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>cell_type_level_2</th>\n",
       "      <th>cell_type_level_3</th>\n",
       "      <th>tumor_original_seurat</th>\n",
       "      <th>malignant_exclude</th>\n",
       "      <th>final_cell_types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>wu01_AAACCTGCACATAACC-1</th>\n",
       "      <td>wu01</td>\n",
       "      <td>11822.0</td>\n",
       "      <td>3295</td>\n",
       "      <td>3377.0</td>\n",
       "      <td>1314</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>In-house: EUS-FNB</td>\n",
       "      <td>wu01</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>False</td>\n",
       "      <td>Monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wu01_AAACCTGCACCACGTG-1</th>\n",
       "      <td>wu01</td>\n",
       "      <td>469.0</td>\n",
       "      <td>293</td>\n",
       "      <td>2235.0</td>\n",
       "      <td>450</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>In-house: EUS-FNB</td>\n",
       "      <td>wu01</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>False</td>\n",
       "      <td>Monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wu01_AAACCTGTCTAACTTC-1</th>\n",
       "      <td>wu01</td>\n",
       "      <td>482.0</td>\n",
       "      <td>243</td>\n",
       "      <td>2436.0</td>\n",
       "      <td>400</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>In-house: EUS-FNB</td>\n",
       "      <td>wu01</td>\n",
       "      <td>Malignant_0</td>\n",
       "      <td>Malignant_0</td>\n",
       "      <td>Malignant_0</td>\n",
       "      <td>False</td>\n",
       "      <td>Malignant - Classical Low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wu01_AAACGGGCAAGGCTCC-1</th>\n",
       "      <td>wu01</td>\n",
       "      <td>1516.0</td>\n",
       "      <td>717</td>\n",
       "      <td>2944.0</td>\n",
       "      <td>717</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>In-house: EUS-FNB</td>\n",
       "      <td>wu01</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>False</td>\n",
       "      <td>Monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wu01_AAACGGGTCATTCACT-1</th>\n",
       "      <td>wu01</td>\n",
       "      <td>876.0</td>\n",
       "      <td>453</td>\n",
       "      <td>2491.0</td>\n",
       "      <td>524</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>In-house: EUS-FNB</td>\n",
       "      <td>wu01</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>Monocyte</td>\n",
       "      <td>False</td>\n",
       "      <td>Monocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTCTACCAGA-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>8237.0</td>\n",
       "      <td>2326</td>\n",
       "      <td>9379.0</td>\n",
       "      <td>2326</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>CAF</td>\n",
       "      <td>CAF</td>\n",
       "      <td>Chan-Seng-Yue</td>\n",
       "      <td>G9903</td>\n",
       "      <td>CAF</td>\n",
       "      <td>CAF</td>\n",
       "      <td>CAF</td>\n",
       "      <td>False</td>\n",
       "      <td>CAF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAGTTGTCGT-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>6484.0</td>\n",
       "      <td>1875</td>\n",
       "      <td>9150.0</td>\n",
       "      <td>1886</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>CAF</td>\n",
       "      <td>CAF</td>\n",
       "      <td>Chan-Seng-Yue</td>\n",
       "      <td>G9903</td>\n",
       "      <td>CAF</td>\n",
       "      <td>CAF</td>\n",
       "      <td>CAF</td>\n",
       "      <td>False</td>\n",
       "      <td>CAF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACAACTTGAC-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>8367.0</td>\n",
       "      <td>2286</td>\n",
       "      <td>9400.0</td>\n",
       "      <td>2286</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>CAF</td>\n",
       "      <td>CAF</td>\n",
       "      <td>Chan-Seng-Yue</td>\n",
       "      <td>G9903</td>\n",
       "      <td>CAF</td>\n",
       "      <td>CAF</td>\n",
       "      <td>CAF</td>\n",
       "      <td>False</td>\n",
       "      <td>CAF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATCCGAACGC-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>20314.0</td>\n",
       "      <td>3975</td>\n",
       "      <td>10684.0</td>\n",
       "      <td>3464</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>Chan-Seng-Yue</td>\n",
       "      <td>G9903</td>\n",
       "      <td>Malignant_0</td>\n",
       "      <td>Malignant_0</td>\n",
       "      <td>Malignant_0</td>\n",
       "      <td>False</td>\n",
       "      <td>Malignant - Classical Low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATCTCTTGAT-1</th>\n",
       "      <td>G9903</td>\n",
       "      <td>7159.0</td>\n",
       "      <td>2162</td>\n",
       "      <td>9260.0</td>\n",
       "      <td>2169</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>Malignant</td>\n",
       "      <td>Chan-Seng-Yue</td>\n",
       "      <td>G9903</td>\n",
       "      <td>Malignant_1</td>\n",
       "      <td>Malignant_1</td>\n",
       "      <td>Malignant_1</td>\n",
       "      <td>False</td>\n",
       "      <td>Malignant - Classical High</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>140612 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   orig.ident  nCount_RNA  \\\n",
       "wu01_AAACCTGCACATAACC-1                                  wu01     11822.0   \n",
       "wu01_AAACCTGCACCACGTG-1                                  wu01       469.0   \n",
       "wu01_AAACCTGTCTAACTTC-1                                  wu01       482.0   \n",
       "wu01_AAACGGGCAAGGCTCC-1                                  wu01      1516.0   \n",
       "wu01_AAACGGGTCATTCACT-1                                  wu01       876.0   \n",
       "...                                                       ...         ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...      G9903      8237.0   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...      G9903      6484.0   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...      G9903      8367.0   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...      G9903     20314.0   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...      G9903      7159.0   \n",
       "\n",
       "                                                    nFeature_RNA  nCount_SCT  \\\n",
       "wu01_AAACCTGCACATAACC-1                                     3295      3377.0   \n",
       "wu01_AAACCTGCACCACGTG-1                                      293      2235.0   \n",
       "wu01_AAACCTGTCTAACTTC-1                                      243      2436.0   \n",
       "wu01_AAACGGGCAAGGCTCC-1                                      717      2944.0   \n",
       "wu01_AAACGGGTCATTCACT-1                                      453      2491.0   \n",
       "...                                                          ...         ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...          2326      9379.0   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...          1875      9150.0   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...          2286      9400.0   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...          3975     10684.0   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...          2162      9260.0   \n",
       "\n",
       "                                                    nFeature_SCT  \\\n",
       "wu01_AAACCTGCACATAACC-1                                     1314   \n",
       "wu01_AAACCTGCACCACGTG-1                                      450   \n",
       "wu01_AAACCTGTCTAACTTC-1                                      400   \n",
       "wu01_AAACGGGCAAGGCTCC-1                                      717   \n",
       "wu01_AAACGGGTCATTCACT-1                                      524   \n",
       "...                                                          ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...          2326   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...          1886   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...          2286   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...          3464   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...          2169   \n",
       "\n",
       "                                                             CELL   CONDITION  \\\n",
       "wu01_AAACCTGCACATAACC-1                             NA_character_ -2147483648   \n",
       "wu01_AAACCTGCACCACGTG-1                             NA_character_ -2147483648   \n",
       "wu01_AAACCTGTCTAACTTC-1                             NA_character_ -2147483648   \n",
       "wu01_AAACGGGCAAGGCTCC-1                             NA_character_ -2147483648   \n",
       "wu01_AAACGGGTCATTCACT-1                             NA_character_ -2147483648   \n",
       "...                                                           ...         ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...  NA_character_ -2147483648   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...  NA_character_ -2147483648   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...  NA_character_ -2147483648   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...  NA_character_ -2147483648   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...  NA_character_ -2147483648   \n",
       "\n",
       "                                                          Patient        Type  \\\n",
       "wu01_AAACCTGCACATAACC-1                             NA_character_ -2147483648   \n",
       "wu01_AAACCTGCACCACGTG-1                             NA_character_ -2147483648   \n",
       "wu01_AAACCTGTCTAACTTC-1                             NA_character_ -2147483648   \n",
       "wu01_AAACGGGCAAGGCTCC-1                             NA_character_ -2147483648   \n",
       "wu01_AAACGGGTCATTCACT-1                             NA_character_ -2147483648   \n",
       "...                                                           ...         ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...  NA_character_ -2147483648   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...  NA_character_ -2147483648   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...  NA_character_ -2147483648   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...  NA_character_ -2147483648   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...  NA_character_ -2147483648   \n",
       "\n",
       "                                                        Cell_type  ...  \\\n",
       "wu01_AAACCTGCACATAACC-1                             NA_character_  ...   \n",
       "wu01_AAACCTGCACCACGTG-1                             NA_character_  ...   \n",
       "wu01_AAACCTGTCTAACTTC-1                             NA_character_  ...   \n",
       "wu01_AAACGGGCAAGGCTCC-1                             NA_character_  ...   \n",
       "wu01_AAACGGGTCATTCACT-1                             NA_character_  ...   \n",
       "...                                                           ...  ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...  NA_character_  ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...  NA_character_  ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...  NA_character_  ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...  NA_character_  ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...  NA_character_  ...   \n",
       "\n",
       "                                                   integrated_snn_res.0.5  \\\n",
       "wu01_AAACCTGCACATAACC-1                                                14   \n",
       "wu01_AAACCTGCACCACGTG-1                                                14   \n",
       "wu01_AAACCTGTCTAACTTC-1                                                 0   \n",
       "wu01_AAACGGGCAAGGCTCC-1                                                 2   \n",
       "wu01_AAACGGGTCATTCACT-1                                                14   \n",
       "...                                                                   ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...                      1   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...                      1   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...                      1   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...                      0   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...                      0   \n",
       "\n",
       "                                                   active.ident  cell_type  \\\n",
       "wu01_AAACCTGCACATAACC-1                                Monocyte   Monocyte   \n",
       "wu01_AAACCTGCACCACGTG-1                                Monocyte   Monocyte   \n",
       "wu01_AAACCTGTCTAACTTC-1                               Malignant  Malignant   \n",
       "wu01_AAACGGGCAAGGCTCC-1                                Monocyte   Monocyte   \n",
       "wu01_AAACGGGTCATTCACT-1                                Monocyte   Monocyte   \n",
       "...                                                         ...        ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...          CAF        CAF   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...          CAF        CAF   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...          CAF        CAF   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...    Malignant  Malignant   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...    Malignant  Malignant   \n",
       "\n",
       "                                                              dataset  \\\n",
       "wu01_AAACCTGCACATAACC-1                             In-house: EUS-FNB   \n",
       "wu01_AAACCTGCACCACGTG-1                             In-house: EUS-FNB   \n",
       "wu01_AAACCTGTCTAACTTC-1                             In-house: EUS-FNB   \n",
       "wu01_AAACGGGCAAGGCTCC-1                             In-house: EUS-FNB   \n",
       "wu01_AAACGGGTCATTCACT-1                             In-house: EUS-FNB   \n",
       "...                                                               ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...      Chan-Seng-Yue   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...      Chan-Seng-Yue   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...      Chan-Seng-Yue   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...      Chan-Seng-Yue   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...      Chan-Seng-Yue   \n",
       "\n",
       "                                                   sample_id  \\\n",
       "wu01_AAACCTGCACATAACC-1                                 wu01   \n",
       "wu01_AAACCTGCACCACGTG-1                                 wu01   \n",
       "wu01_AAACCTGTCTAACTTC-1                                 wu01   \n",
       "wu01_AAACGGGCAAGGCTCC-1                                 wu01   \n",
       "wu01_AAACGGGTCATTCACT-1                                 wu01   \n",
       "...                                                      ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...     G9903   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...     G9903   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...     G9903   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...     G9903   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...     G9903   \n",
       "\n",
       "                                                   cell_type_level_2  \\\n",
       "wu01_AAACCTGCACATAACC-1                                     Monocyte   \n",
       "wu01_AAACCTGCACCACGTG-1                                     Monocyte   \n",
       "wu01_AAACCTGTCTAACTTC-1                                  Malignant_0   \n",
       "wu01_AAACGGGCAAGGCTCC-1                                     Monocyte   \n",
       "wu01_AAACGGGTCATTCACT-1                                     Monocyte   \n",
       "...                                                              ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...               CAF   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...               CAF   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...               CAF   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...       Malignant_0   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...       Malignant_1   \n",
       "\n",
       "                                                   cell_type_level_3  \\\n",
       "wu01_AAACCTGCACATAACC-1                                     Monocyte   \n",
       "wu01_AAACCTGCACCACGTG-1                                     Monocyte   \n",
       "wu01_AAACCTGTCTAACTTC-1                                  Malignant_0   \n",
       "wu01_AAACGGGCAAGGCTCC-1                                     Monocyte   \n",
       "wu01_AAACGGGTCATTCACT-1                                     Monocyte   \n",
       "...                                                              ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...               CAF   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...               CAF   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...               CAF   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...       Malignant_0   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...       Malignant_1   \n",
       "\n",
       "                                                   tumor_original_seurat  \\\n",
       "wu01_AAACCTGCACATAACC-1                                         Monocyte   \n",
       "wu01_AAACCTGCACCACGTG-1                                         Monocyte   \n",
       "wu01_AAACCTGTCTAACTTC-1                                      Malignant_0   \n",
       "wu01_AAACGGGCAAGGCTCC-1                                         Monocyte   \n",
       "wu01_AAACGGGTCATTCACT-1                                         Monocyte   \n",
       "...                                                                  ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...                   CAF   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...                   CAF   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...                   CAF   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...           Malignant_0   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...           Malignant_1   \n",
       "\n",
       "                                                   malignant_exclude  \\\n",
       "wu01_AAACCTGCACATAACC-1                                        False   \n",
       "wu01_AAACCTGCACCACGTG-1                                        False   \n",
       "wu01_AAACCTGTCTAACTTC-1                                        False   \n",
       "wu01_AAACGGGCAAGGCTCC-1                                        False   \n",
       "wu01_AAACGGGTCATTCACT-1                                        False   \n",
       "...                                                              ...   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...             False   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...             False   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...             False   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...             False   \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...             False   \n",
       "\n",
       "                                                              final_cell_types  \n",
       "wu01_AAACCTGCACATAACC-1                                               Monocyte  \n",
       "wu01_AAACCTGCACCACGTG-1                                               Monocyte  \n",
       "wu01_AAACCTGTCTAACTTC-1                              Malignant - Classical Low  \n",
       "wu01_AAACGGGCAAGGCTCC-1                                               Monocyte  \n",
       "wu01_AAACGGGTCATTCACT-1                                               Monocyte  \n",
       "...                                                                        ...  \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGGTTTC...                         CAF  \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCAAG...                         CAF  \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCACA...                         CAF  \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...   Malignant - Classical Low  \n",
       "G9903_filtered_gene_bc_matrices_h5.h5_TTTGTCATC...  Malignant - Classical High  \n",
       "\n",
       "[140612 rows x 25 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3d8a618b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<140612x30840 sparse matrix of type '<class 'numpy.float32'>'\n",
       "\twith 295800072 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a.layers['counts']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2dd4e43",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
